feat(examples): add disperse (v2)

agherasie opened the pull request:
This PR adds a gno version of the [disperse ethereum app](https://disperse.app/) to gno.land !
Another attempt was made in https://github.com/gnolang/gno/pull/1414 but we have decided to pick up from @leohhhn 's work with @lennyvong and made this PR, tested using txtar tests to avoid the situation described in https://github.com/gnolang/gno/issues/2595
There is also an older (but functional for the most part) version [deployed in test4](https://test4.gno.land/r/g1w62226g8hykfmtuasvz80rdf0jl6phgxsphh5v/testing/disperse2?help) with a linked [webapp](https://gno-disperse.netlify.app/)

<img width="649" alt="image" src="https://github.com/user-attachments/assets/ad26a1a9-0447-4333-858a-253441c457ba">

<details><summary>Contributors' checklist...</summary>

- [X] Added new tests, or not needed, or not feasible
- [X] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [X] Updated the official documentation or not needed
- [X] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [X] Added references to related issues and PRs
- [X] Provided any useful hints for running manual tests
- [ ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

moul commented:
It looks quite interesting. Please note that I'm working on a GRC20 refactor that could impact this PR. Feel free to merge it, and I can fix your example in my OR. Alternatively, you can wait until I complete the refactor. #2551

gfanton commented on examples/gno.land/r/demo/disperse/disperse.gno:
```diff
@@ -0,0 +1,128 @@
+func DisperseToken(addresses []std.Address, coins std.Coins) {
+	caller := std.GetOrigCaller() // get tx sender
+
+	// Check if caller correctly sent positive amount of coins
+	for _, coin := range coins {
+		if coin.Amount <= 0 {
+			panic(ErrNegativeCoinAmount)
+		}
+	}
+
+	// Check if number of addresses and coins match
+	if len(addresses) != len(coins) {
+		panic(ErrNumAddrValMismatch)
+	}
```
consider placing this logic in a dedicated method, as the same logic is used in the previous method

agherasie replied:
https://github.com/gnolang/gno/pull/2613/commits/7437f607c6c942090a6398d46a30d024d9b17b44

gfanton commented on examples/gno.land/r/demo/disperse/disperse.gno:
```diff
@@ -0,0 +1,128 @@
+	}
+
+	// Disperse tokens
+	for i, address := range addresses {
+		for y, coin := range coins {
+			if y == i {
+				tokens.Transfer(coin.Denom, address, uint64(coin.Amount))
+			}
+		}
+	}
+}
```
since we assume `len(addresses) == len(coins)`, you could probably optimize this:
```suggestion
	for i := 0; i < len(address); i++ {
			tokens.Transfer(coins[i].Denom, address[i], uint64(coins[i].Amount))
		}
	}
}
```

gfanton commented on gno.land/cmd/gnoland/testdata/disperse.txtar:
Huh, I'm not sure using txtar to do this kind of test is a good idea. txtar is meant for CLI testing or reproducible bugs, at last for now. 
Did you check some `_filetest.gno` files to see if you can reproduce what you want using the `// SEND` pattern ?

agherasie replied:
https://github.com/gnolang/gno/pull/2613/commits/6418040c7439d6176042aca0211fc61497a5e235

Thanks for the suggestion, filetests worked really well !

gfanton commented on gno.land/cmd/gnoland/testdata/disperse.txtar:
```diff
@@ -0,0 +1,131 @@
+### Execute disperse gnot
+
+# Verifying default balance
+gnokey query bank/balances/g1jg8mtutu9khhfwc4nxmuhcpftf0pajdhfvsqf5
```
use `adduser <user>` command, then use `$USER_ADDR_<user>` to grab the address

gfanton reviewed (changes requested):
Looks really good, love the idea! 
(check comments for (minor) requested changes)

leohhhn commented on examples/gno.land/r/demo/disperse/disperse.gno:
```diff
@@ -0,0 +1,122 @@
+
+	// Send coins
+	for i, _ := range addresses {
+		banker.SendCoins(realmAddr, addresses[i], std.Coins{coins[i]})
```
You can use `std.NewCoins()`
```suggestion
		banker.SendCoins(realmAddr, addresses[i], std.NewCoins(coins[i]))
```

leohhhn commented on examples/gno.land/r/demo/disperse/disperse.gno:
```diff
@@ -0,0 +1,122 @@
+	}
+}
+
+// DisperseGnot parses receivers and amounts and sends out gnot
```
From what I see, it also returns any accidentally sent coins. Maybe it would be good to document this, or remove this functionality altogether

agherasie replied:
it is present in the [original ethereum smart contract](https://etherscan.io/address/0xD152f549545093347A162Dce210e7293f1452150#code#L1) so we thought it would be best to keep it

leohhhn commented on examples/gno.land/r/demo/disperse/disperse.gno:
```diff
@@ -0,0 +1,122 @@
+	}
+}
+
+// DisperseToken disperses tokens to multiple addresses
```
What kind of token exactly? Maybe it would be better to have it called DisperseGRC20, or similarly?

agherasie replied:
https://github.com/gnolang/gno/pull/2613/commits/042ec411f405c1e18227a7731159b80f5f7ab5b3

leohhhn commented on examples/gno.land/r/demo/disperse/disperse.gno:
```diff
@@ -0,0 +1,122 @@
+	}
+
+	// Check if the realm has zero ugnot
+	if banker.GetCoins(realmAddr).AmountOf("ugnot") < coinSent.AmountOf("ugnot") {
```
Is this case possible? If it is, maybe the check should be higher in code, to stop execution right away, since it is a critical error?

leohhhn replied:
Did you have any updates to this convo?

agherasie replied:
sorry, i didn't provide a commit link because i switched that logic up a bunch of times throughout several commits
I ended up changing the check and placing it at the beginning of the function like you said

leohhhn commented on examples/gno.land/r/demo/disperse/disperse.gno:
```diff
@@ -0,0 +1,122 @@
+}
+
+// DisperseToken disperses tokens to multiple addresses
+func DisperseToken(addresses []std.Address, coins std.Coins) {
```
Interesting.

While it might be tempting to use std.Coins to represent GRC20 tokens, I don't think we should use it because it will create confusion on what is actually being sent (coins or grc20?). Let's switch to using the symbol + amount for GRC20. 
@moul, what do you think about this?


Also, please document that this function can only be called after approving disperse for the tokens you want to send.

agherasie replied:
https://github.com/gnolang/gno/pull/2613/commits/dcd81fe1c62445f8924ea0b07716eb68adb079b9
https://github.com/gnolang/gno/pull/2613/commits/56d07169fe92cd10363165713edf9f0f88ba4958

leohhhn commented on examples/gno.land/r/demo/disperse/disperse.gno:
```diff
@@ -0,0 +1,122 @@
+
+// DisperseToken disperses tokens to multiple addresses
+func DisperseToken(addresses []std.Address, coins std.Coins) {
+	caller := std.GetOrigCaller() // get tx sender
```
```suggestion
	caller := std.PrevRealm().Addr() // get tx sender
```

leohhhn commented on examples/gno.land/r/demo/disperse/util.gno:
```diff
@@ -0,0 +1,62 @@
+		addr := std.Address(str)
+		if !addr.IsValid() {
+			return nil, ErrInvalidAddress
+		}
```
In Go/Gno, its good practice to add a new line after a block, like this: 
```suggestion
		}
		
```

Could you apply this style across the board in the app?

agherasie replied:
https://github.com/gnolang/gno/pull/2613/commits/0a87b66f6037f03a688651a62589e878f1af161b

leohhhn reviewed (approved):
LGTM. Thanks for continuing this!

Left some comments to be addressed before we can go ahead with the merge 💯

agherasie commented on examples/gno.land/r/demo/disperse/errors.gno:
```diff
@@ -0,0 +1,13 @@
+	ErrNegativeCoinAmount       = errors.New("coin amount cannot be negative")
+	ErrBalanceNotEqToAmountSent = errors.New("balance needs to be equal to amount sent")
+	ErrArgLenAndSentLenMismatch = errors.New("mismatch between coins sent and args called")
+	ErrWrongAmount              = errors.New("wrong coin amount")
```
decided to remove this because it is a redundant check

agherasie commented on examples/gno.land/r/demo/disperse/disperse.gno:
```diff
@@ -0,0 +1,122 @@
+package disperse
```
https://github.com/gnolang/gno/pull/2613/commits/d91b227305b9dbc33b8d2b4bdeb3b716c952546f

agherasie commented on examples/gno.land/r/demo/disperse/errors.gno:
```diff
@@ -0,0 +1,13 @@
+import "errors"
+
+var (
+	ErrNotEnoughCoin            = errors.New("not enough coin sent in")
```
https://github.com/gnolang/gno/pull/2613/commits/5fd031e767718344020c2b2000d043727c847109

agherasie commented on examples/gno.land/r/demo/disperse/disperse.gno:
```diff
@@ -0,0 +1,122 @@
+	checkNumAddrValMatch(addresses, coins)
+
+	// Get address of Disperse realm
+	realmAddr := std.CurrentRealm().Addr()
```
https://github.com/gnolang/gno/pull/2613/commits/7f8aa8485594bb3489678ee59173f6fcd9afeec8

agherasie commented on examples/gno.land/r/demo/disperse/disperse.gno:
```diff
@@ -0,0 +1,122 @@
+	}
+}
+
+// DisperseGnot parses receivers and amounts and sends out gnot
```
https://github.com/gnolang/gno/pull/2613/commits/d41e44cb3b67eec14f3c489adcbfb5c641396f28

agherasie commented:
> LGTM. Thanks for continuing this!
> 
> Left some comments to be addressed before we can go ahead with the merge 💯

@leohhhn I've addressed all of your comments and sent the relevant commits in replies except a few ones which were split across several commits

leohhhn commented:
Hey @agherasie, I left some more comments that should be addressed, but the gist is there.

agherasie commented on examples/gno.land/r/demo/disperse/disperse.gno:
```diff
@@ -0,0 +1,108 @@
+		}
+	}
+
+	var totalAmount int64
+	for _, coin := range coins {
+		totalAmount += coin.Amount
+	}
```
it was before but not anymore you're right !

agherasie commented on examples/gno.land/r/demo/disperse/disperse.gno:
```diff
@@ -0,0 +1,108 @@
+
+	for i := 0; i < len(addresses); i++ {
+		// Transfer tokens into the realm
+		tokens.TransferFrom(symbols[i], caller, std.CurrentRealm().Addr(), amounts[i])
```
yeah I thought so, we kept it because it was done in the original implementation in solidity :
```sol
function disperseToken(IERC20 token, address[] recipients, uint256[] values) external {
        uint256 total = 0;
        for (uint256 i = 0; i < recipients.length; i++)
            total += values[i];
        require(token.transferFrom(msg.sender, address(this), total));
        for (i = 0; i < recipients.length; i++)
            require(token.transfer(recipients[i], values[i]));
    }
```
is this also the case for ethereum ? or am I missing something that differs between the implementations themselves ?

as for the fix :
https://github.com/gnolang/gno/pull/2613/commits/d05632a7711b9629469fc981cf295631ac0a9a57

leohhhn replied:
Hmm, maybe it's due to an old version of ERC20. From what I saw with the current implementation, `transferFrom` in Solidity ([OZ](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/0b58a783b9b33b63eef2994af8958c0c6a72dc51/contracts/token/ERC20/ERC20.sol#L190)) reverts on insufficient balance, but that might not be the case for the 0.4.X version which Disperse is made with.

If the tests pass, lets keep it only with `transferFrom` 👍

agherasie commented on examples/gno.land/r/demo/disperse/disperse.gno:
```diff
@@ -0,0 +1,108 @@
+	for _, coin := range coinSent {
+		leftoverAmt := banker.GetCoins(realmAddr).AmountOf(coin.Denom)
+		if leftoverAmt > 0 {
+			send := std.Coins{{coin.Denom, leftoverAmt}}
```
https://github.com/gnolang/gno/pull/2613/commits/286da8304a9846f54fcf6fcab35b66d433bcc1db

agherasie commented on examples/gno.land/r/demo/disperse/disperse.gno:
```diff
@@ -0,0 +1,108 @@
+
+	coins := make(std.Coins, len(parsedAmounts))
+	for i, amount := range parsedAmounts {
+		coins[i] = std.Coin{"ugnot", amount}
```
https://github.com/gnolang/gno/pull/2613/commits/286da8304a9846f54fcf6fcab35b66d433bcc1db

agherasie commented on examples/gno.land/r/demo/disperse/disperse.gno:
```diff
@@ -0,0 +1,108 @@
+	banker := std.GetBanker(std.BankerTypeOrigSend)
+
+	// Check if realm balance is equal to amount sent
+	if banker.GetCoins(realmAddr).AmountOf("ugnot") != coinSent.AmountOf("ugnot") {
```
https://github.com/gnolang/gno/pull/2613/commits/128a8e87132ac05b4759e8c8369301ea642f9d06

agherasie commented on examples/gno.land/r/demo/disperse/disperse.gno:
```diff
@@ -0,0 +1,108 @@
+
+	// Return possible leftover coins
+	for _, coin := range coinSent {
+		leftoverAmt := banker.GetCoins(realmAddr).AmountOf(coin.Denom)
```
I can't really do this because of the new changes in https://github.com/gnolang/gno/pull/2613/commits/128a8e87132ac05b4759e8c8369301ea642f9d06
Since I would get a `ErrMismatchBetweenSentAndParams` error because I'm manipulating the coin amounts mid loop
